Initialize AbstractTestQueryFramework lazily

findepi opened the pull request:
Previously the `AbstractTestQueryFramework` was initialized with
`@BeforeClass`. This was a clean approach but led to surefire+TestNG
problem https://issues.apache.org/jira/browse/SUREFIRE-1967: multiple
test instances being initialized and kept around while only two (or
whatever test parallelism is) are actually needed. This in turn led to
tests potentially exhausting available memory and failing with OOM.

This PR tries to workaround this problem by introducing lazy
initialization in `AbstractTestQueryFramework`. It has not been verified
whether this solves the problem. Also, timing of class tear-down wasn't
verified either.

findepi commented on testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueryFramework.java:
```diff
@@ -81,20 +80,33 @@
-    @BeforeClass
-    public void init()
-            throws Exception
+    // TODO (https://github.com/trinodb/trino/issues/11294) make this back as `@BeforeClass init()`
```
https://github.com/trinodb/trino/issues/11294

losipiuk commented on testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueryFramework.java:
```diff
@@ -83,18 +82,34 @@
+    // TODO (https://github.com/trinodb/trino/issues/11294) make this back as `@BeforeClass public void init()`
+    //   - remove all explicit callers
+    //   - remove synchronization
+    protected synchronized void initIfNeeded()
```
feels like this should be `private` as subclasses cannot control the call-lifecycle of this method (cannot access `initialized` variable)

findepi replied:
it needs to be available to subclasses

- let's say `createQueryRunner` allocates some resources (eg docker container)
- and a subclass's test method uses that container (before using the query runner)

for this to work, the subclass needs to invoke `initIfNeeded` explicitly.

(yes, this is a breaking change)

losipiuk replied:
Ok so the assumption is that `initIfNeeded` would not be overridden - just called by the subclasses. And any subclass-specific initialization goes to `createQueryRunner` I guess the structure makes sense then.

findepi replied:
> Ok so the assumption is that `initIfNeeded` would not be overridden

yes! i made it `final` to convey that.

findepi commented on testing/trino-testing/src/main/java/io/trino/testing/AbstractTestQueryFramework.java:
```diff
@@ -81,20 +80,36 @@
-        h2QueryRunner = afterClassCloser.register(new H2QueryRunner());
-        sqlParser = new SqlParser();
-        queryAssertions = new io.trino.sql.query.QueryAssertions(queryRunner);
+        if (initialized) {
```
`Suppliers.memoize` might be useful for single field, here we have 3.

also, i check on the field being null in some places, so Supplier wouldn't really work

findepi commented:
i added a `@BeforeMethod` to trigger initialization to avoid subclasses  having to invoke `initIfNeeded`.
there will still be cases where this is needed -- subclasses with `@BeforeClass` -- but hopefully minimal

however, `@BeforeMethod` is generally forbidden in a context of  a concurrent. This particular case is safe though (synchronized & idempotent), so i added `ReportMultiThreadedBeforeOrAfterMethod.Suppress` annotation.

findepi commented:
@hashhar suggested that simply switching to `air.test.parallel=instances` may be enough of a solution (i.e. until we have surefire with https://issues.apache.org/jira/browse/SUREFIRE-1967 in use).

@hashhar @findinpath let me know your feedback from the testing

findepi commented:
converted to draft. May be unnecessary after https://github.com/trinodb/trino/pull/12192 .
